sim/hunter/focus_fire.go

```diff
@@ -0,0 +1,101 @@
+package hunter
+
+import (
+	"time"
+
+	"github.com/wowsims/sod/sim/core"
+	"github.com/wowsims/sod/sim/core/proto"
+)
+
+func (hunter *Hunter) registerFocusFireSpell() {
+	if hunter.pet == nil || !hunter.HasRune(proto.HunterRune_RuneBracersFocusFire) {
+		return
+	}
+
+	focusFireMetrics := hunter.pet.Metrics.NewResourceMetrics(core.ActionID{SpellID: 428726}, proto.ResourceType_ResourceTypeEnergy)
+	focusFireActionId := core.ActionID{SpellID: 428726}
+	focusFireFrenzyActionId := core.ActionID{SpellID: 428728}
+
+	// For tracking in timeline
+	hunterFrenzyAura := hunter.RegisterAura(core.Aura{
+		Label:     "Focus Fire Frenzy (Hunter)",
+		ActionID:  focusFireFrenzyActionId,
+		Duration:  time.Second * 10,
+		MaxStacks: 5,
+	})
+
+	hunterPetFrenzyAura := hunter.pet.RegisterAura(core.Aura{
+		Label:     "Focus Fire Frenzy",
+		ActionID:  focusFireFrenzyActionId,
+		Duration:  time.Second * 10,
+		MaxStacks: 5,
+		OnStacksChange: func(aura *core.Aura, sim *core.Simulation, oldStacks int32, newStacks int32) {
+			// revert previous attack speed bonus from stacks and apply new attack speed bonus
+			aura.Unit.MultiplyMeleeSpeed(sim, 1/(1+(0.06*float64(oldStacks))))
+			aura.Unit.MultiplyMeleeSpeed(sim, 1+(0.06*float64(newStacks)))
+
+  			// this is just so we can see the frenzy stacks working on the sim timeline
+			if !hunterFrenzyAura.IsActive() { hunterFrenzyAura.Activate(sim) }
+			hunterFrenzyAura.SetStacks(sim, newStacks)
+		},
+	})
+
+	hunterFocusFireAura := hunter.RegisterAura(core.Aura{
+		Label:    "Focus Fire",
+		ActionID: focusFireActionId,
+		Duration: time.Second * 20,
+		OnGain: func(aura *core.Aura, sim *core.Simulation) {
+			frenzyStacks := hunterPetFrenzyAura.GetStacks()
+			aura.Unit.MultiplyRangedSpeed(sim, 1+(0.03*float64(frenzyStacks)))
+		},
+		OnExpire: func(aura *core.Aura, sim *core.Simulation) {
+			frenzyStacks := hunterPetFrenzyAura.GetStacks()
+			aura.Unit.MultiplyRangedSpeed(sim, 1/(1+(0.03*float64(frenzyStacks))))
+		},
```

Comment:
the stacks on the pet aura can change between the OnGain/OnExpire (example OnGain with 1 stack, OnExpire with 5) which make stats "leak". You should setup this aura with its own stacks and use OnStacksChange the same way you do it for the pet frenzy aura. Then all you need to do is set the stacks to the current amount the pet frenzy has and the OnStacksChange will properly take care of the rest.

Reply:
I totally missed this, I updated to use your solution passing in the stack amount from the spell itself